src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -0,0 +1,513 @@
+            var originalCodeBlockSpan = codeBlockRange.AsTextSpan(originalText);
+
+            // Sometimes the C# formatter edits outside the range we supply. Filter out those edits.
+            var changes = edits.Select(e => e.AsTextChange(originalText)).Where(c => originalCodeBlockSpan.Contains(c.Span)).ToArray();
+            if (changes.Length == 0)
+            {
+                return originalText;
+            }
+
+            // Apply the C# edits to the document.
+            var changedText = originalText.WithChanges(changes);
+            TrackChangeInSpan(originalText, originalCodeBlockSpan, changedText, out var changedCodeBlockSpan, out var changeEncompassingSpan);
+
+            // We now have the changed document with C# edits. But it might be indented more/less than what we want depending on the context.
+            // So, we want to bring each line to the right level of indentation based on where the block is in the document.
+            // We also need to only do this for the lines that are part of the input range to respect range formatting.
+            var desiredIndentationLevel = context.Indentations[(int)codeBlockRange.Start.Line].IndentationLevel + 1;
+            var editsToApply = new List<TextChange>();
+            var inputSpan = context.Range.AsTextSpan(originalText);
+            TrackChangeInSpan(originalText, inputSpan, changedText, out var changedInputSpan, out _);
+            var changedInputRange = changedInputSpan.AsRange(changedText);
+
+            for (var i = (int)changedInputRange.Start.Line; i <= changedInputRange.End.Line; i++)
+            {
+                var line = changedText.Lines[i];
+                if (line.Span.Length == 0)
+                {
+                    // Empty line. C# formatter didn't remove it so we won't either.
+                    continue;
+                }
+
+                if (!changedCodeBlockSpan.Contains(line.Start))
+                {
+                    // Defensive check to make sure we're not handling lines that are not part of the current code block.
+                    continue;
+                }
+
+                var leadingWhitespace = line.GetLeadingWhitespace();
+                var minCSharpIndentLength = context.Options.TabSize * minCSharpIndentLevel;
+                if (leadingWhitespace.Length < minCSharpIndentLength)
+                {
+                    // For whatever reason, the C# formatter decided to not indent this. Leave it as is.
+                    continue;
+                }
+                else
+                {
+                    var effectiveDesiredIndentationLevel = desiredIndentationLevel - minCSharpIndentLevel;
+                    if (effectiveDesiredIndentationLevel < 0)
+                    {
+                        // This means that we need to unindent.
+                        var length = Math.Abs(effectiveDesiredIndentationLevel * context.Options.TabSize);
```

Comment:
This breaks with tabs instead of spaces doesn't it?

Reply:
Why so?

Reply:
Same answer as above.